Switch back to rouge gem for synthax highlighting

rumpelsepp opened the pull request:
I have mainly created the rugments fork for the purpose of improving
gitlab's highlighting. Nowadays IMO it works way better than the old
highlight.js solution. But the development is stuck on my side because
of a couple of _personal_ reasons:
- I have finished my studies; last months I was writing my master
  thesis. So there was a huge time problem. I am sorry for that.
- I had to move to Munich due to getting a (paid) job. Searching a
  flat here is horrible... :)
- Last but not least, maintaining the same code base in two seperate
  projects is a mess.

I have decided to switch back to rouge due to several reasons:
- In the beginning I was quite motivated, but since I start
  working on my new job next week, the best solution IMO is
  switching back to upstream rouge.
- Rouge is continously improving:
  https://github.com/jneen/rouge/blob/master/CHANGELOG.md
  http://rouge.jneen.net/
- I am going to submit pull request to rouge, though. I had some
  ideas how it can be improved.
- There should be absolutely no regressions with this change. Most
  likely this pull request will almost fix some minor bugs.
- One less gem in gitlab is a good thing. since Gitlab is quite a huge
  bundle of gems. Reducing complexity should be a major milestone.

Work to be done:
- I am going to submit a pull request to rouge which includes the
  gitlab html formatter.
- The gollum-libs gem needs to be updated due to this pull request:
  https://github.com/gollum/gollum-lib/pull/159

jneen commented:
:+1:

tsigo commented:
:+1:

randx commented:
:+1:

tsigo commented:
How's this going?

rumpelsepp commented:
As long as gollum-lib has not done a release which includes my commit https://github.com/gollum/gollum-lib/commit/9c042ea0914a7714abbda9d89e234c8a79fba893, this rull request is blocked I think.

rumpelsepp committed:
Add HTMLGitlab formatter

This custom formatter for rouge is needed to generate HTML output
specifically for gitlab. Since its usecase is mostly suitable for
gitlab it had been rejected upstream:
https://github.com/jneen/rouge/pull/268

Thanks a lot to @stanhu, @jneen and @tsigo for review!

rumpelsepp committed:
Replace Rugments with Rouge

I have mainly created the rugments fork for the purpose of improving
gitlab's highlighting. Nowadays IMO it works way better than the old
highlight.js solution. But the development is stuck on my side because
of a couple of personal reasons:

  * I have finished my studies; last months I was writing my master
    thesis. So there was a huge time problem. I am sorry for that.

  * I had to move to Munich due to getting a (paid) job. Searching a
    flat here is horrible... :)

  * Last but not least, maintaining the same code base in two seperate
    projects is a mess.

I have decided to switch back to rouge due to several reasons:

  * In the beginning I was quite motivated, but since I start
    working on my new job next week, the best solution IMO is
    switching back to upstream rouge.

  * Rouge is continously improving:
    https://github.com/jneen/rouge/blob/master/CHANGELOG.md
    http://rouge.jneen.net/

  * There should be absolutely no regressions with this change. Most
    likely this pull request will almost fix some minor bugs.

  * One less gem in gitlab is a good thing. since Gitlab is quite a
    huge bundle of gems. Reducing complexity should be a major
    milestone.

Thanks a lot to @stanhu and @jneen for the review!

rumpelsepp commented:
Should be ready to merge. I have splitted the PR into several commits and bumped rouge's version manually. Hopefully, the tests will pass. :)

rumpelsepp commented:
Tests fail because this commit https://github.com/jneen/rouge/commit/03b03c598b6108294d3671988bc97f27ce4f9897 is not yet released by rouge upstream. :(

rumpelsepp commented:
I have asked @jneen to bump the rouge version to 1.9.1: https://github.com/jneen/rouge/issues/283

rumpelsepp commented:
could somebody restart the tests?

rumpelsepp commented:
oh sry, i habe to update to rouge 1.9.1 first :)

jneen commented:
Hi everyone, I highly recommend not making yourselves dependent on rouge releases if yall are planning on contributing code to rouge.  This is because I don't have infinite time to cut releases after every pull request.  If you want edge features, bundler is quite happy to pull in a gem from a git hash, and that's the approach I recommend.

randx commented:
@jneen I understand how release take your time. Thank you for releasing 1.9.1. We would use git hash but it means rely on github.com which is not good for us.  Thank you for `rouge` !

jneen commented:
Ok! If yall want to mirror rouge on gitlab that'd be ok with me too.

rumpelsepp commented:
@randx I think I will update this pull request and resolve merge conflicts tonight.

stanhu commented:
@rumpelsepp Great! It looks like we may still be blocked by gollum updates?

rumpelsepp commented:
There is this patch: https://github.com/rumpelsepp/gitlabhq/commit/fe15e2671ea3ee7baea4fb57b554622dfc1eafd7

I have to bump it to rouge 1.9.1. It is not pretty, but it should work, since the gollum creater has accepted the pull request everything should be fine.

rumpelsepp committed:
Manually update rouge to 1.9.1

Since gollum-lib blocks rouge from updating, let's bump the version
manually. The fix has already been accepted upstream:
https://github.com/gollum/gollum-lib/commit/9c042ea0914a7714abbda9d89e234c8a79fba893

rumpelsepp commented:
Should be ready! Can anyone restart the tests? This one is not caused by my pull request I think:
```
DETAIL:  Process 3372 waits for AccessExclusiveLock on relation 16751 of database 16509; blocked by process 3597.
Process 3597 waits for AccessShareLock on relation 16612 of database 16509; blocked by process 3372.
HINT:  See server log for query details.
```
https://semaphoreci.com/gitlabhq/gitlabhq/branches/pull-request-9341/builds/7

rumpelsepp commented:
@tsigo: I restarted the build manually (by adding a comma in the commit message); now the build still fails with:

```
Error summary:
  Errors (1)
    Project Source Git Blame :: I blame file :: And I click Blame button
```
https://semaphoreci.com/gitlabhq/gitlabhq/branches/pull-request-9341/builds/8

Since I am using rouge 1.9.1 this should not happen… Any ideas?

stanhu commented:
The error is:

```
 undefined method `html_safe' for nil:NilClass
        /home/runner/gitlabhq/app/helpers/blob_helper.rb:15:in `rescue in highlight'
```

Somehow the return value of the GitLab formatter `format` function is `nil`?

rumpelsepp commented:
Thanks! that's seems the problem. I renamed it to `render`: https://github.com/rumpelsepp/gitlabhq/commit/123b3633f57ba4f20550f2b488979e1b1df0e5c0#diff-3958226e9c9e3f5c381a5596c8ad8687R53

stanhu commented:
Does that break the subclassing of Rouge's `Formatter` class?

rumpelsepp commented:
No, I was too fast with my answer. There is an
``` ruby
alias_method :format, :render
```
below.

stanhu commented:
Should the `format_untableized` and `format_tableized` end with `html` to ensure the `unless` blocks don't cause the functions to return `nil`?

tsigo commented:
> Should format_untableized and format_tableized end with html to ensure the unless blocks don't cause the functions to return nil?

Yes, good call.

rumpelsepp commented:
Thanks! I wouldn't have found that in a thousand years... Nice catch. So hopefully the tests are happy now. :)

randx commented:
@rumpelsepp what you doing is awesome! :+1:

rumpelsepp commented:
Thank you very much! :)
Another issue spotted by the test suite. :/

stanhu commented:
@rumpelsepp `blob_helper_spec.rb` is failing because the merge dropped the `continue` parameter:

```diff
-  def highlight(blob_name, blob_content, nowrap: false, continue: false)
+  def highlight(blob_name, blob_content, nowrap = false)
```

```diff
-      result = @formatter.format(@lexer.lex(blob_content, continue: continue)).html_safe
+      result = @formatter.format(@lexer.lex(blob_content)).html_safe
```

stanhu commented on lib/rouge/formatters/html_gitlab.rb:
```diff
@@ -0,0 +1,169 @@
+        @lineanchors = lineanchors
+        @lineanchorsid = lineanchorsid
+        @anchorlinenos = anchorlinenos
+        @inline_theme = inline_theme
```
I suggest instantiating the theme here as rouge does:

https://github.com/jneen/rouge/blob/master/lib/rouge/formatters/html.rb#L33

stanhu commented on lib/rouge/formatters/html_gitlab.rb:
```diff
@@ -0,0 +1,169 @@
+          val
+        else
+          if @inline_theme
+            theme = Theme.find(@inline_theme).new
```
This will cause a new Theme to be instantiated for each `span` call. Unnecessary overhead.

stanhu commented on app/helpers/emails_helper.rb:
```diff
@@ -31,8 +31,8 @@ def action_title(url)
   def color_email_diff(diffcontent)
-    formatter = Rugments::Formatters::HTML.new(cssclass: "highlight", inline_theme: :github)
-    lexer = Rugments::Lexers::Diff.new
+    formatter = Rouge::Formatters::HTML.new(cssclass: "highlight", inline_theme: :github)
```
It looks like `html_inline.rb` doesn't expect a string as the theme. As a workaround, this seems to make things work:

```ruby
    formatter = Rouge::Formatters::HTML.new(cssclass: "highlight", inline_theme: Rouge::Themes::Github.new)
```

But I'm confused because while I find `html_inline.rb` in the .gem file, it doesn't appear to be in the [source](https://github.com/jneen/rouge/tree/master/lib/rouge/formatters). Is this auto-generated, @jneen?

jneen replied:
I'm a bit confused - my branch with HTMLInline hasn't been merged yet - unless you have a similar file in the gitlab project?

jneen replied:
Either way, the parameter here is misspelled - it should be `css_class: 'highlight'`.

stanhu replied:
Thanks for the review, @jneen. I downloaded `rouge-1.9.1.gem` from RubyGems, ran `gem unpack` on it, and then did this:

```
$ find . -name html_inline.rb
./rouge-1.9.1/lib/rouge/formatters/html_inline.rb
```

Is this some packaging issue?

jneen replied:
ugh, yes it's probably a packaging issue.  is it causing problems for it to be there?

stanhu replied:
I don't think so. It just led to confusion because I was trying to debug a failed test, and my debugger pointed me towards `html_inline.rb` instead of `html.rb`. Apparently, the use of the symbol `:github` as opposed to `github` caused this error.

stanhu commented:
Looks like we're close, @rumpelsepp! Just a few minor changes here, and tests should pass.

rumpelsepp commented:
Thanks for reviewing this! I should buy you guys a beer I think... :) I am not sure of I could finish these tasks tonight; it has been a long day at work. :)

rumpelsepp commented:
Thanks a lot for the reviews! An updated patch series is online; the tests are running :sunglasses:

stanhu commented:
Thanks! The tests may fail due to an unrelated failure that I'm trying to patch now. :)

rumpelsepp commented:
Ok. Failed now; could you retrigger the tests when you are done? :)

stanhu commented:
@rumpelsepp Kicked off a new build. Not sure if you rebased master recently, but if you did you may need to do it again.

stanhu commented on app/helpers/blob_helper.rb:
```diff
@@ -1,6 +1,6 @@
 module BlobHelper
-  def highlight(blob_name, blob_content, nowrap: false, continue: false)
-    @formatter ||= Rugments::Formatters::HTML.new(
+  def highlight(blob_name, blob_content, nowrap = false, continue: false)
```
Looks like a typo here: `nowrap: false`, not `nowrap = false`.

rumpelsepp replied:
thx; fixed.

rumpelsepp commented:
Thanks, I hope the tests passes now. :tada:

rumpelsepp commented:
Hmm. I'll fix that last issue tomorrow or so. I'm in bed now and coding from here is kinda difficult :-)

stanhu commented on app/helpers/blob_helper.rb:
```diff
@@ -8,11 +8,11 @@ def highlight(blob_name, blob_content, nowrap: false, continue: false)
 
     begin
-      @lexer ||= Rugments::Lexer.guess(filename: blob_name, source: blob_content).new
+      @lexer ||= Rouge::Lexer.guess(filename: blob_name, source: blob_content)
```
You forgot to include the `continue` parameter here.

stanhu replied:
The test fails here because the constructor of `Lexer` calls `new(opts).lex(stream, &b)`, but `opts` is not actually used in the `lex` call. You can work around this issue by adding `.new` at the end of the line.

rumpelsepp replied:
Ahh. Nice catch, thx.

stanhu commented on app/helpers/emails_helper.rb:
```diff
@@ -31,8 +31,8 @@ def action_title(url)
   def color_email_diff(diffcontent)
-    formatter = Rugments::Formatters::HTML.new(cssclass: "highlight", inline_theme: :github)
-    lexer = Rugments::Lexers::Diff.new
+    formatter = Rouge::Formatters::HTML.new(css_class: 'highlight', inline_theme: :github)
```
See the discussion earlier. `:github` doesn't work; you have to use `'github'`.

rumpelsepp commented:
hmm, not sure why it still fails with: 

```
rspec ./spec/helpers/blob_helper_spec.rb:22 # BlobHelper highlight should highlight continued blocks
```

The `continue` argument is present again (I think It got lost due to rebasing in the first place). Any ideas?

stanhu commented:
Yes, see my comment above.

stanhu commented:
@rumpelsepp Merged with pleasure. Thanks!

rumpelsepp commented:
... and thank you very much for the review!

tsigo commented:
@rumpelsepp Thank you for your work on this.

@stanhu Thanks for taking over the review.
